src/security/token/TKTokenCryptoEngine.m

```diff
@@ -48,15 +53,74 @@ - (TKSignature *)signData:(NSData *)data
             usingKeyLevel:(Key_Level)keyLevel
                    reason:(NSString *)reason
                   onError:(OnError)onError {
+    if (!useLocalAuthentication) {
+        return [self createSignature:data usingKeyLevel:keyLevel];
+    }
+    
+    if (keyLevel >= Key_Level_Low) {
+        // We don't check DeviceOwnerAuthentication for Key_Level_Low
+        return [self createSignature:data usingKeyLevel:keyLevel];
+    }
+    
+    LAContext *context = [[LAContext alloc] init];
+    
+    if (![context canEvaluatePolicy:LAPolicyDeviceOwnerAuthentication error:nil]) {
+        // If device don't support Device Owner Authentication, skip as success
+        return [self createSignature:data usingKeyLevel:keyLevel];
+    }
+    
+    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+    __block TKSignature *signature = nil;
+    __block NSError *laError = nil;
+    
+    [context evaluatePolicy:LAPolicyDeviceOwnerAuthentication
+            localizedReason:reason
+                      reply:^(BOOL success, NSError *error) {
+                          // This is a backend thread
+                          if (success) {
+                              signature = [self createSignature:data usingKeyLevel:keyLevel];
+                          }
+                          else {
+                              laError = error;
+                          }
+                          dispatch_semaphore_signal(semaphore);
+                      }];
+    
+    dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
```

Comment:
This blocks the main thread, right? Shouldn't we do all of the stuff below in the callback?

Reply:
This is blocking the main thread, but we shall not do the below stuff in the callback.
this local authentication call back is in non-main thread, while we definitely want to keep the onError call on main  thread. 
And starting a main thread block in the call back before semaphore finish is a bad idea, since we need to run the runloop manually. So the best practice is to cache the result and do it after the semaphore.

Reply:
Ok, we should marshall it to the main thread then. But we should not block the thread. This freezes everything, doesn't it?